homeassistant/components/onvif/camera.py

```diff
@@ -176,7 +177,9 @@ async def async_added_to_hass(self):
         """Run when entity about to be added to hass."""
         uri_no_auth = await self.device.async_get_stream_uri(self.profile)
         self._stream_uri = uri_no_auth.replace(
-            "rtsp://", f"rtsp://{self.device.username}:{self.device.password}@", 1
+            "rtsp://",
+            f"rtsp://{urllib.parse.quote(self.device.username)}:{urllib.parse.quote(self.device.password)}@",
```

Comment:
We should not solve this using string replacement/encoding. Instead, we should parse the URL to re-construct it. E.g., using YARL.